---
title: "Blast virus NT report"
date: "`r Sys.Date()`"
params:
  rmd: "taxonomy_report.Rmd"
output: 
  html_document:
    theme: united
    highlight: tango
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)
```

```{r}
library(tidyverse)
library(DT)
library(formattable)
input <- unique(file.path("..", snakemake@input))
```

## Sample `r snakemake@params[[1]]`

### Number of unique queries per tax_id
**In case of ties (e-value), first top hits were selected**.

```{r}
taxonomy <- grep("names.csv|division.csv", input, value = TRUE)
results <- setdiff(input, taxonomy)
qc <- grep("fastq_join|bwa_mem", input, value = TRUE)
results <- setdiff(results, qc)
names <- read_csv(grep("names.csv", taxonomy, value = TRUE))
division <- read_csv(grep("division.csv", taxonomy, value = TRUE))
known <- data_frame(path = unlist(results)) %>% 
  mutate(data = map(path, read_csv)) %>% 
  unnest()
scientific_names <- filter(names, name_class == "scientific name")
known_names <- left_join(known, scientific_names)
known_names <- left_join(known_names, division)
first_top_hits <- known_names %>% 
  group_by(query) %>% 
  filter(rank(Hsp_evalue, ties.method = "first") == 1) %>% 
  ungroup()
```

```{r, fig.cap="Number of sequence reads (unique queries) per tax_id."}
seq_reads <- first_top_hits %>%
  group_by(tax_id, name_txt, div = division_name) %>%
  summarise(N = n(),
            `%` = percent(N / nrow(.))) %>%
  ungroup() %>%
  arrange(desc(N)) %>% 
  formattable()
seq_reads %>% 
  as.datatable(caption = paste0("Number of sequence reads (unique queries) per tax_id. Number of unique tax_id-s is ", nrow(seq_reads),"."), rownames = FALSE, filter = "top", options = list(autoWidth = TRUE))
```

Drop queries and get distinct gi-s:
```{r}
by_gi <- first_top_hits %>%
  group_by(gi, tax_id, name_txt) %>% 
 summarise(N = n(),
            `%` = percent(N / nrow(.))) %>%
  ungroup() %>%
  arrange(desc(N)) %>% 
  formattable()
by_gi %>% 
  as.datatable(caption = paste0("Number of tax_id per top hit gi (unique top hit). Number of unique top hits ", nrow(by_gi), "."), rownames = FALSE, filter = "top", options = list(autoWidth = TRUE))
```

### Number of phage and other virus hits

Approximate number of tax_id-s belonging to virus or bacteriophage class. 
```{r}
by_division <- first_top_hits %>%
  group_by(division_id) %>%
  summarise(N = n(),
            `%` = percent(N / nrow(.))) %>%
  ungroup() %>%
  arrange(desc(N))
by_division %>% 
  left_join(division) %>% 
  dplyr::select(id = division_id, cde = division_cde, name = division_name, N, `%`) %>% 
  formattable() %>%  
  as.datatable(rownames = FALSE, caption = "Number of sequence reads belonging to virus or bacteriophage classes. Unclassified taxa are shown by name.")
```

### Source
<a download="taxonomy_report.Rmd" href="`r base64enc::dataURI(file = params$rmd, mime = 'text/rmd', encoding = 'base64')`">R Markdown source file (to produce this document)</a>

